lib/FreeRTOS/freertos-plus-mqtt/agent_command_pool.c

```diff
@@ -0,0 +1,165 @@
+
+/* Header include. */
+#include "agent_command_pool.h"
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The pool of command structures used to hold information on commands (such
+ * as PUBLISH or SUBSCRIBE) between the command being created by an API call and
+ * completion of the command by the execution of the command's callback.
+ */
+static Command_t commandStructurePool[ MQTT_COMMAND_CONTEXTS_POOL_SIZE ];
+
+/**
+ * @brief A counting semaphore used to guard the pool of Command_t structures.  To
+ * obtain a structure first decrement the semaphore count.  To return a structure
+ * increment the semaphore count after the structure is back in the pool.
+ */
+static SemaphoreHandle_t freeCommandStructMutex = NULL;
+
+static volatile bool initializedSemaphore = false;
+
+/*-----------------------------------------------------------*/
+
+static void initializePool()
+{
+    bool destroySemaphore = true;
+    SemaphoreHandle_t tempMutex = xSemaphoreCreateCounting( MQTT_COMMAND_CONTEXTS_POOL_SIZE, MQTT_COMMAND_CONTEXTS_POOL_SIZE );
+
+    taskENTER_CRITICAL();
+    {
+        if( !initializedSemaphore )
+        {
+            memset( ( void * ) commandStructurePool, 0x00, sizeof( commandStructurePool ) );
+            freeCommandStructMutex = tempMutex;
+            destroySemaphore = false;
+        }
+
+        initializedSemaphore = true;
+    }
+    taskEXIT_CRITICAL();
+
+    /* Destroy the semaphore if it was not used. */
+    if( destroySemaphore )
+    {
+        vSemaphoreDelete( tempMutex );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+Command_t * Agent_GetCommand( uint32_t blockTimeMs )
+{
+    Command_t * structToUse = NULL;
+    size_t i;
+
+    /* Check here so we do not enter a critical section every time. */
+    if( !initializedSemaphore )
+    {
+        initializePool();
+        configASSERT( freeCommandStructMutex ); /*_RB_ Create all objects here statically. */
+    }
+
+    /* Check counting semaphore has been created. */
+    if( freeCommandStructMutex != NULL )
+    {
+        /* If the semaphore count is not zero then a command context is available. */
+        if( xSemaphoreTake( freeCommandStructMutex, pdMS_TO_TICKS( blockTimeMs ) ) == pdPASS )
+        {
+            for( i = 0; i < MQTT_COMMAND_CONTEXTS_POOL_SIZE; i++ )
+            {
+                taskENTER_CRITICAL();
+                {
+                    /* If the commandType is NONE then the structure is not in use. */
+                    if( commandStructurePool[ i ].commandType == NONE )
+                    {
+                        LogDebug( ( "Removed Command Context %d from pool", ( int ) i ) );
+                        structToUse = &( commandStructurePool[ i ] );
+
+                        /* To show the struct is no longer available to be returned
+                         * by calls to Agent_ReleaseCommand(). */
+                        structToUse->commandType = !NONE;
```

Comment:
`!NONE`  ick

Reply:
Yeah it isn't pretty but it's what's currently there. I'd rather separate it now into a new file and make incremental improvements rather than fixing everything in this PR